File: Python/2816. Double a Number Represented as a Linked List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    carry = 0

    def doubleIt(self, head: Optional[ListNode]) -> Optional[ListNode]:
        def dfs(node):
            if node.next:
                dfs(node.next)

            temp = node.val * 2 + self.carry

            node.val = temp % 10
            self.carry = temp // 10

        dfs(head)

        # Check if carry
        if self.carry != 0:
            new = ListNode(
                val=self.carry,
                next=head,
            )
        else:
            new = head

        return new

File: C/2816. Double a Number Represented as a Linked List.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
int carry;

void dfs(struct ListNode* node) {
    if (node->next)
        dfs(node->next);

    int temp = node->val * 2 + carry;

    node->val = temp % 10;
    carry = (int)(temp / 10);
}

struct ListNode* doubleIt(struct ListNode* head) {
    carry = 0;

    dfs(head);

    if (carry) {
        struct ListNode* new = malloc(sizeof(struct ListNode));

        new-> val = carry;
        new->next = head;

        return new;
    }

    return head;
}
